arch/x86/core/pcie.c

```diff
@@ -270,20 +251,23 @@ uint8_t arch_pcie_msi_vectors_allocate(unsigned int priority,
 #endif /* CONFIG_INTEL_VTD_ICTL */
 
 		for (i = 0; i < n_vector; i++) {
-			uint32_t data;
-
-			data = _read_pcie_irq_data(vectors[i].bdf);
-
-			vectors[i].arch.irq = pcie_alloc_irq(vectors[i].bdf);
-
-			_write_pcie_irq_data(vectors[i].bdf, data);
 
 			vectors[i].arch.vector =
 				z_x86_allocate_vector(priority, prev_vector);
 			if (vectors[i].arch.vector < 0) {
 				return 0;
 			}
 
+#ifdef CONFIG_PCIE_MSI_X
+			if (!vectors[0].msix)
+#endif
+			{
+				if (prev_vector != -1) {
+					__ASSERT(prev_vector == (vectors[i].arch.vector - 1),
```

Comment:
This is exactly why VT-D was selected for generating multi-vector MSI. Zephyr's vector allocator is not bullet proof on contiguous allocation. And you really cannot fail here like that since this will happen very often. Let's take ehl's ethernet stuff: they require 32 vectors (on which only 16 are actually used, 15 at the beginning and the last one at the end. Don't ask me why they went for this weird setup, in addition of using msi instead of msi-x...), and there can be 2 device instances, and more devices may require such feature (who knows what devices ppl will plug in). Depending on priority and initialization order, the vector array is soon going to be like an emmental cheese: too many holes and no places to fit such long contiguous vectors request.

Note that this is the same issue on Linux: it cannot allocate contiguous vectors and make use of VT-D for MSI multi-vector purpose.

Reply:
This is the limitation of MSI by design and is addressed by introducing MSI-X in later PCI spec. So for device exposing MSI capability, system SW have two choice either allocate vector as power of two requested by MSI cap register or emulate as MSI-X. We have to keep this 1st option currently (which is also by design correct only) and we don't have 2nd working option currently available.

Reply:
Remove that change altogether, since next commit is going to fix it properly.